affinity/cpp-20/d0796r1.md

```diff
@@ -255,6 +255,10 @@ An interesting question which arises here is whether the system topology of an e
 
 Note that this is different from devices that go online or offline during execution: The devices themselves are online, they have not been found (or used) by the program until the appropriate discovery stage has been executed.
 
+There are generally two main reasons why you would want the **system resource topology** to be dynamic after the initial point of discovery. Firstly for fault tolerance, being able to handle a resource becoming unavailable at some point after the initial discovery while the application is running. Secondly for detecting a new resource, being able to discover new resources added to the system at some point after the initial discovery while the application is running. Though the former case is much more common than the latter.
+
+MPI originally (in MPI-1) did not support dynamic process management and fault tolerance, all processes which were capable of communicating with each other would be identified and fixed during initialisation. PVM (which was originally designed for heterogeneous compute systems), used an alternative execution model of manually spawning processes from the main process and with it allowed for fault tolerance. This demonstrated that it could be implemented and could be useful, which lead to MPI also introducing dynamic process management and fault tolerance in MPI-2. However, althogh these features are available, the execution model of having all processes fixed on initialisation is generally still the prefered approach for HPC aplications. This execution model is shared by SHMEM, Fortran coarrays and UPC++.
```

Comment:
From a historic perspective, different programming models have tackled the problem of dynamic system resources following various approaches. In particular, PVM  (parallel virtual machine, XXX) enabled resources to be discovered during runtime since its conception. This lead to MPI to add it in later revisions. 
However, as far as we know, the dynamic process feature is not widely used in HPC environments.
Other programming models for HPC environments support a fixed set of processors on initialization library time, such as SHMEM, Fortran coarrays and UPC++.
There is, however, plenty of work on these models to address fault tolerance, in particular, MPI has some fault-tolerance support among others.

Reply:
MPI's fault tolerance isn't really part of the standard yet -- at least it didn't make it into 3.0, despite quite a lot of discussion.

Reply:
Good point, edited my suggested changes to reflect it.

Reply:
@Ruyk @mhoemmen I've revised this paragraph to reflect your suggestions. I also made the point that whilst MPI current doesn't natively support fault tolerance, it can be implemented on top or provided via extensions, I hope I have understood this correctly.